### PR TITLE
Ports Beestation's radio uplink

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
A port of https://github.com/BeeStation/BeeStation-Hornet/pull/7363

The radio uplink is now reworked to require saying a codeword over the secret .d channel. This means that you can do it hands free, but people around you can hear you whispering the codeword
## Why It's Good For The Game
Current radio uplink is dumb, it has a lot of drawbacks (losing common radio, fiddling with the tgui slider, having to reset the frequency after) that its almost unusable. This new uplink has a more unique trigger method with its own ups and downs
## Changelog
:cl:
add: New radio uplink! Simply speak your codeword into the :d channel to unlock it
code: new COMSIG_RADIO_NEW_MESSAGE signal
/:cl:

### CODE REVIEW

File: code/__DEFINES/radio.dm [hunk 1]
```diff
@@ -56,6 +60,7 @@
 // Frequencies are always odd numbers and range from 1201 to 1599.
 
 #define FREQ_SYNDICATE 1213  // Nuke op comms frequency, dark brown
+#define FREQ_UPLINK 1214	// Dummy loopback frequency, used for radio uplink. Not seen in game.
```
Comment (ZephyrTFA) [inline code/__DEFINES/radio.dm:63]: Frequencies are odd

File: code/datums/mind/antag.dm [hunk 2]
```diff
@@ -126,6 +126,11 @@
 	var/implant = FALSE
 
 	var/uplink_spawn_location = traitor_mob.client?.prefs?.read_preference(/datum/preference/choiced/uplink_location)
+	var/cant_speak = if(HAS_TRAIT(traitor_mob, TRAIT_MUTE) || traitor_mob.mind?.assigned_role.title == JOB_MIME)
```
Comment (ZephyrTFA) [inline code/datums/mind/antag.dm:129]:
```suggestion
	var/cant_speak = (HAS_TRAIT(traitor_mob, TRAIT_MUTE) || traitor_mob.mind?.assigned_role.title == JOB_MIME)
```

File: code/datums/mind/antag.dm [hunk 3]
```diff
@@ -164,7 +169,7 @@
 	new_uplink.uplink_handler.assigned_role = traitor_mob.mind.assigned_role.title
 	new_uplink.uplink_handler.assigned_species = traitor_mob.dna.species.id
 	if(uplink_loc == R)
-		unlock_text = "Your Uplink is cunningly disguised as your [R.name]. Simply dial the frequency [format_frequency(new_uplink.unlock_code)] to unlock its hidden features."
+		unlock_text = "Your Uplink is cunningly disguised as your [R.name]. Simply speak [new_uplink.unlock_code] into frequency :d to unlock its hidden features."
```
Comment (ZephyrTFA) [inline code/datums/mind/antag.dm:172]:
```suggestion
		unlock_text = "Your Uplink is cunningly disguised as your [R.name]. Simply speak \"[new_uplink.unlock_code]\" into frequency :d to unlock its hidden features."
```

Comment (ZephyrTFA) [reply code/datums/mind/antag.dm:172]: Also `frequency :d` seems wrong

File: code/game/objects/items/devices/radio/radio.dm [hunk 4]
```diff
@@ -224,6 +224,8 @@
 /obj/item/radio/talk_into(atom/movable/talking_movable, message, channel, list/spans, datum/language/language, list/message_mods)
 	if(SEND_SIGNAL(talking_movable, COMSIG_MOVABLE_USING_RADIO, src) & COMPONENT_CANNOT_USE_RADIO)
 		return
+	if(SEND_SIGNAL(src, COMSIG_RADIO_NEW_MESSAGE, M, message, channel) & COMPONENT_CANNOT_USE_RADIO)
```
Comment (ZephyrTFA) [inline code/game/objects/items/devices/radio/radio.dm:227]:
```suggestion
	if(SEND_SIGNAL(src, COMSIG_RADIO_NEW_MESSAGE, talking_movable, message, channel) & COMPONENT_CANNOT_USE_RADIO)
```
Stuff like this is why you need to run the liner before making a PR. Don't just blindly copy and paste

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/radio/radio.dm (hunk 4 above)
Comment (etherware-novice) [reply code/game/objects/items/devices/radio/radio.dm:227]: to be fair i wasnt using that variable for anything anyways

**Comment by SomeoneYouProbablyKnow** [comment]
Very neat. Might actually make it better than the mighty pen. But that's good, long time players are too good at finding uplinks on people.

File: code/datums/components/uplink.dm [hunk 5]
```diff
@@ -359,6 +359,22 @@
 	if(ismob(master.loc))
 		interact(null, master.loc)
 
+/datum/component/uplink/proc/new_message(datum/source, user, message, channel)
+	SIGNAL_HANDLER
+
+	if(channel != RADIO_CHANNEL_UPLINK)
+		return
+
+	. = COMPONENT_CANNOT_USE_RADIO // we're using uplink channel, cut all communication
```
Comment (Mothblocks) [inline code/datums/components/uplink.dm:368]: return COMPONENT_CANNOT_USE_RADIO in relevant places instead.

https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md#dot-variable-

File: code/datums/components/uplink.dm [hunk 6]
```diff
@@ -385,15 +401,15 @@
...
+		unlock_note = "<B>Radio Passcode:</B> [unlock_code] ([P.name], :d channel)."
 	else if(istype(parent,/obj/item/pen))
 		unlock_note = "<B>Uplink Degrees:</B> [english_list(unlock_code)] ([P.name])."
 
 /datum/component/uplink/proc/generate_code()
 	if(istype(parent,/obj/item/modular_computer/tablet))
 		return "[rand(100,999)] [pick(GLOB.phonetic_alphabet)]"
 	else if(istype(parent,/obj/item/radio))
-		return return_unused_frequency()
+		return "[pick(GLOB.phonetic_alphabet)]"
```
Comment (Mothblocks) [inline code/datums/components/uplink.dm:412]:
```suggestion
		return pick(GLOB.phonetic_alphabet)
```

File: code/datums/mind/antag.dm [hunk 7]
```diff
@@ -126,6 +126,11 @@
 	var/implant = FALSE
 
 	var/uplink_spawn_location = traitor_mob.client?.prefs?.read_preference(/datum/preference/choiced/uplink_location)
+	var/cant_speak = (HAS_TRAIT(traitor_mob, TRAIT_MUTE) || traitor_mob.mind?.assigned_role.title == JOB_MIME)
+	if(uplink_spawn_location == UPLINK_RADIO && cant_speak)
+		if(!silent)
+			to_chat(traitor_mob, span_warning("You have been deemed ineligable for a radio uplink. Supplying standard uplink intead."))
```
Comment (Mothblocks) [inline code/datums/mind/antag.dm:132]:
```suggestion
			to_chat(traitor_mob, span_warning("You have been deemed ineligible for a radio uplink. Supplying standard uplink instead."))
```

File: code/datums/mind/antag.dm [hunk 8]
```diff
@@ -164,7 +169,7 @@
 	new_uplink.uplink_handler.assigned_role = traitor_mob.mind.assigned_role.title
 	new_uplink.uplink_handler.assigned_species = traitor_mob.dna.species.id
 	if(uplink_loc == R)
-		unlock_text = "Your Uplink is cunningly disguised as your [R.name]. Simply dial the frequency [format_frequency(new_uplink.unlock_code)] to unlock its hidden features."
+		unlock_text = "Your Uplink is cunningly disguised as your [R.name]. Simply speak \"[new_uplink.unlock_code]\" into frequency :d to unlock its hidden features."
```
Comment (Mothblocks) [inline code/datums/mind/antag.dm:172]: Is this information in the antag menu somewhere? How will you know if you miss this message?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/mind/antag.dm (hunk 8 above)
Comment (etherware-novice) [reply code/datums/mind/antag.dm:172]:
```
/datum/antagonist/traitor/ui_static_data(mob/user)
	var/datum/component/uplink/uplink = uplink_ref?.resolve()
	var/list/data = list()
...
		data["uplink_unlock_info"] = uplink.unlock_text``` yep, this unlock text is displayed under the antag menu

Comment (Mothblocks) [reply code/datums/mind/antag.dm:172]: Ok cool

Comment (etherware-novice) [reply code/datums/mind/antag.dm:172]: and it should also be under your ic notes, since it mentions :d in there too

**Comment by Mothblocks** [comment]
Undraft if ready

**Pull request merged**